prime-router/src/main/kotlin/tokens/Server2ServerAuthentication.kt

```diff
@@ -35,47 +50,150 @@ class Server2ServerAuthentication : Logging {
         logger.error(msg)
     }
 
-    /**
-     * @return true if jwsString is a validly signed Sender token, false if it is unauthorized
-     * If it is valid, then its ok to move to the next step, then give the sender an Access token.
-     */
-    fun checkSenderToken(
+    private fun parseJwt(jwsString: String, scope: String, metadata: Metadata): ParsedJwt {
+        val workflowEngine = WorkflowEngine.Builder().metadata(metadata).build()
+
+        // parseClaimsJwt will throw an exception if the string includes a signature, even when just parsing
+        // the claims and headers (which are just Base64 encoded).
+        // See https://github.com/jwtk/jjwt/issues/86
+        val i = jwsString.lastIndexOf('.')
+        val withoutSignature = jwsString.substring(0, i + 1)
+        val jwt = Jwts.parserBuilder().build().parseClaimsJwt(withoutSignature)
+
+        val claims = jwt.body
+        val headers = jwt.header
+        val issuer = claims.issuer // client_id
+
+        // The issuer should always be the organization name as this most closely matches the client_id
+        // description from the FHIR spec
+        // http://hl7.org/fhir/uv/bulkdata/authorization/index.html#signature-verification:~:text=Upon%20registration%2C%20the%20client%20SHALL%20be%20assigned%20a%20client_id%2C%20which%20the%20client%20SHALL%20use%20when%20requesting%20an%20access%20token
+        // However, in order to be backwards compatible, the issuer claim can be either the name of the sender
+        // or the name of the organization.
+        var maybeOrganization = workflowEngine.settings.findOrganization(issuer)
+        val maybeSender = workflowEngine.settings.findSender(issuer)
+        if (maybeOrganization == null && maybeSender != null) {
+            maybeOrganization = workflowEngine.settings.findOrganization(maybeSender.organizationName)
+        }
+
+        if (maybeOrganization == null) {
+            throw Server2ServerAuthenticationException("$issuer was not valid.", scope)
+        }
+
+        val maybeKid = headers["kid"] as String?
+        val alg = headers["alg"] as String
+        val kty = KeyType.forAlgorithm(Algorithm.parse(alg))
+
+        return ParsedJwt(maybeOrganization, maybeSender, maybeKid, kty)
+    }
+
+    private fun getPossibleSigningKeys(parsedJwt: ParsedJwt, scope: String): List<Key> {
```

Comment:
These private methods should be unit tested since they are not trivial. If that's difficult to do due to their private nature, I suggest breaking these out into a separate class? Thoughts?

Reply:
I think I'd prefer to keep them private since they really shouldn't be used outside of this code.  I think I can likely just make them internal and test them that way.

I'll also drop the coverage since I think the current tests exercise all the lines.